In [1]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.preprocessing import LabelEncoder

In [2]:
# Import the dataset
training_df = pd.read_excel("/Users/jasoneisele/Desktop/Relay_Data_HW2_2_2_2_2/relaytrain.xlsx")

In [3]:
print("The shape of our training dataframe is {}".format(training_df.shape[0]) + " rows, and {}".format(training_df.shape[1]) + " columns")

The shape of our training dataframe is 24579 rows, and 16 columns


In [4]:
# Check the top 5 rows
training_df.head()

custid  retained    created           firstorder            lastorder  \
0  6H6T6N         0 2012-09-28  2013-08-11 00:00:00  2013-08-11 00:00:00   
1  APCENR         1 2010-12-19  2011-04-01 00:00:00  2014-01-19 00:00:00   
2  7UP6MS         0 2010-10-03  2010-12-01 00:00:00  2011-07-06 00:00:00   
3  7ZEW8G         0 2010-10-22  2011-03-28 00:00:00  2011-03-28 00:00:00   
4  8V726M         1 2010-11-27  2010-11-29 00:00:00  2013-01-28 00:00:00   

   esent   eopenrate  eclickrate  avgorder   ordfreq  paperless  refill  \
0     29  100.000000    3.448276     14.52  0.000000          0       0   
1     95   92.631579   10.526316     83.69  0.181641          1       1   
2      0    0.000000    0.000000     33.58  0.059908          0       0   
3      0    0.000000    0.000000     54.96  0.000000          0       0   
4     30   90.000000   13.333333    111.91  0.008850          0       0   

   doorstep  train     favday city  
0         0      1     Monday  CHO  
1         1      1     Friday  CHO  
2         0      1  Wednesday  CHO  
3         0      1   Thursday  RIC  
4         0      1     Monday  RIC

In [5]:
#Check for missing values
training_df.isnull().sum()

custid        17
retained       0
created       17
firstorder    17
lastorder     17
esent          0
eopenrate      0
eclickrate     0
avgorder       0
ordfreq        0
paperless      0
refill         0
doorstep       0
train          0
favday         0
city           0
dtype: int64

In [6]:
# Drop rows where custid has missing values
training_df = training_df[pd.notnull(training_df['custid'])]

In [7]:
# Recheck for missing values
training_df.isnull().sum()

custid        0
retained      0
created       0
firstorder    0
lastorder     0
esent         0
eopenrate     0
eclickrate    0
avgorder      0
ordfreq       0
paperless     0
refill        0
doorstep      0
train         0
favday        0
city          0
dtype: int64

In [8]:
print("The new shape of our training dataframe is {}".format(training_df.shape[0]) + " rows, and {}".format(training_df.shape[1]) + " columns")

The new shape of our training dataframe is 24562 rows, and 16 columns


### Our dataset had 24579 rows before and 24562 after dropping the missing rows where custid feature was missing a value. 17 rows total ~ 0.069% which is extremely small

In [9]:
print("The data lost was " + str(float((17/24579)*100)) + "%")

The data lost was 0.069164734122625%


In [10]:
# Check the data types of our features
training_df.dtypes

custid                object
retained               int64
created       datetime64[ns]
firstorder            object
lastorder             object
esent                  int64
eopenrate            float64
eclickrate           float64
avgorder             float64
ordfreq              float64
paperless              int64
refill                 int64
doorstep               int64
train                  int64
favday                object
city                  object
dtype: object

In [11]:
# Encode/transform categorical variables 'favday' and 'city'
le = LabelEncoder()
training_df['favday'] = le.fit_transform(training_df['favday'])
training_df['city'] = le.fit_transform(training_df['city'])
training_df.head()

custid  retained    created           firstorder            lastorder  \
0  6H6T6N         0 2012-09-28  2013-08-11 00:00:00  2013-08-11 00:00:00   
1  APCENR         1 2010-12-19  2011-04-01 00:00:00  2014-01-19 00:00:00   
2  7UP6MS         0 2010-10-03  2010-12-01 00:00:00  2011-07-06 00:00:00   
3  7ZEW8G         0 2010-10-22  2011-03-28 00:00:00  2011-03-28 00:00:00   
4  8V726M         1 2010-11-27  2010-11-29 00:00:00  2013-01-28 00:00:00   

   esent   eopenrate  eclickrate  avgorder   ordfreq  paperless  refill  \
0     29  100.000000    3.448276     14.52  0.000000          0       0   
1     95   92.631579   10.526316     83.69  0.181641          1       1   
2      0    0.000000    0.000000     33.58  0.059908          0       0   
3      0    0.000000    0.000000     54.96  0.000000          0       0   
4     30   90.000000   13.333333    111.91  0.008850          0       0   

   doorstep  train  favday  city  
0         0      1       1     1  
1         1      1       0     1  
2         0      1       6     1  
3         0      1       4     3  
4         0      1       1     3

In [12]:
# Instantiate a LogisiticRegression class
lr = LogisticRegression()

In [13]:
# Split training dataset into X_train and y_train
X_train, y_train = (training_df[['esent','eopenrate','eclickrate','avgorder','ordfreq','paperless','refill','doorstep','train','favday','city']],training_df['retained'])

In [14]:
# Check the first 5 rows of X_train
X_train.head()

esent   eopenrate  eclickrate  avgorder   ordfreq  paperless  refill  \
0     29  100.000000    3.448276     14.52  0.000000          0       0   
1     95   92.631579   10.526316     83.69  0.181641          1       1   
2      0    0.000000    0.000000     33.58  0.059908          0       0   
3      0    0.000000    0.000000     54.96  0.000000          0       0   
4     30   90.000000   13.333333    111.91  0.008850          0       0   

   doorstep  train  favday  city  
0         0      1       1     1  
1         1      1       0     1  
2         0      1       6     1  
3         0      1       4     3  
4         0      1       1     3

In [15]:
# Check the first 5 rows of y_train
y_train.head()

0    0
1    1
2    0
3    0
4    1
Name: retained, dtype: int64

In [16]:
# Fit our Logistic Regression model to the training dataset
logreg = lr.fit(X_train, y_train)
print("The coefficients of our logreg model are " + str(list(logreg.coef_)))
print("The intercept of our logreg model is " + str(logreg.intercept_))

The coefficients of our logreg model are [array([ 0.21207411,  0.00579462,  0.01315318, -0.00355594, -0.57737349,
        0.38339248,  0.74983522,  0.80515318, -1.14030893, -0.03492538,
       -0.10563622])]
The intercept of our logreg model is [-1.14030893]


In [17]:
print("The accuracy of our Logistic Regression model is " + str(logreg.score(X_train, y_train)*100) + "%")

The accuracy of our Logistic Regression model is 94.18206986401759%


In [18]:
# Import our testing dataset
test_df = pd.read_excel("/Users/jasoneisele/Desktop/Relay_Data_HW2_2_2_2_2/relaytest.xlsx")

In [19]:
# Check the first 5 rows
test_df.head()

custid  retained    created           firstorder            lastorder  \
0  QPRVDD         1 2010-11-09  2011-05-12 00:00:00  2013-10-21 00:00:00   
1  HUBE83         1 2011-02-01  2011-02-01 00:00:00  2013-10-27 00:00:00   
2  GW8AUL         1 2010-09-24  2011-05-05 00:00:00  2014-01-19 00:00:00   
3  WYSPZ7         0 2010-09-06  2011-03-22 00:00:00  2014-01-14 00:00:00   
4  ZLK255         0 2010-08-22  2011-09-06 00:00:00  2011-09-06 00:00:00   

   esent  eopenrate  eclickrate  avgorder   ordfreq  paperless  refill  \
0    174  96.551724   12.643678     94.50  0.047032          1       1   
1     49  20.408163    2.040816     43.21  0.016016          0       0   
2     51  64.705882   25.490196     81.15  0.151515          1       0   
3      9  44.444444    0.000000    213.52  0.150632          1       1   
4      0   0.000000    0.000000    146.98  0.000000          0       0   

   doorstep  train     favday city  
0         0      0     Friday  CHO  
1         0      0     Monday  CHO  
2         0      0     Monday  RIC  
3         0      0  Wednesday  CHO  
4         0      0    Tuesday  CHO

In [20]:
print("The shape of our test dataframe is {}".format(test_df.shape[0]) + " rows, and {}".format(test_df.shape[1]) + " columns")

The shape of our test dataframe is 6222 rows, and 16 columns


In [21]:
#Check for any missing values
test_df.isnull().sum()

custid        3
retained      0
created       3
firstorder    3
lastorder     3
esent         0
eopenrate     0
eclickrate    0
avgorder      0
ordfreq       0
paperless     0
refill        0
doorstep      0
train         0
favday        0
city          0
dtype: int64

In [22]:
# Drop rows where custid has missing values
test_df = test_df[pd.notnull(test_df['custid'])]

In [23]:
# Recheck for missing values
test_df.isnull().sum()

custid        0
retained      0
created       0
firstorder    0
lastorder     0
esent         0
eopenrate     0
eclickrate    0
avgorder      0
ordfreq       0
paperless     0
refill        0
doorstep      0
train         0
favday        0
city          0
dtype: int64

In [24]:
# Encode categorical variables favday and city
le = LabelEncoder()
test_df['favday'] = le.fit_transform(test_df['favday'])
test_df['city'] = le.fit_transform(test_df['city'])
test_df.head()

custid  retained    created           firstorder            lastorder  \
0  QPRVDD         1 2010-11-09  2011-05-12 00:00:00  2013-10-21 00:00:00   
1  HUBE83         1 2011-02-01  2011-02-01 00:00:00  2013-10-27 00:00:00   
2  GW8AUL         1 2010-09-24  2011-05-05 00:00:00  2014-01-19 00:00:00   
3  WYSPZ7         0 2010-09-06  2011-03-22 00:00:00  2014-01-14 00:00:00   
4  ZLK255         0 2010-08-22  2011-09-06 00:00:00  2011-09-06 00:00:00   

   esent  eopenrate  eclickrate  avgorder   ordfreq  paperless  refill  \
0    174  96.551724   12.643678     94.50  0.047032          1       1   
1     49  20.408163    2.040816     43.21  0.016016          0       0   
2     51  64.705882   25.490196     81.15  0.151515          1       0   
3      9  44.444444    0.000000    213.52  0.150632          1       1   
4      0   0.000000    0.000000    146.98  0.000000          0       0   

   doorstep  train  favday  city  
0         0      0       0     1  
1         0      0       1     1  
2         0      0       1     3  
3         0      0       6     1  
4         0      0       5     1

In [25]:
# Split our testing dataset into X_test and y_test
X_test, y_test = (test_df[['esent','eopenrate','eclickrate','avgorder','ordfreq','paperless','refill','doorstep','train','favday','city']],test_df['retained'])

In [26]:
# Take our same fitted Logistic Regression model and use it to create a new predicted feature 'y_pred'
y_pred = logreg.predict(X_test)

In [27]:
# Compare the precision, recal, and f1-score between our predicted values y_pred) and our real test values (y_test)
print(classification_report(y_test, y_pred))

             precision    recall  f1-score   support

          0       0.99      0.69      0.81      1286
          1       0.93      1.00      0.96      4933

avg / total       0.94      0.93      0.93      6219



In [28]:
# Confusion matrix shows the hit rate
print(confusion_matrix(y_test, y_pred))

[[ 887  399]
 [   8 4925]]


In [29]:
print("The accuracy of our Logistic Regression model on the test data is " + str(logreg.score(X_test, y_test)*100) + "%")

The accuracy of our Logistic Regression model on the test data is 93.45553947579997%


### Our model only lost a 0.7% dip in accuracy from training data to test data

### Question 1
Use the Relay train data to develop a model to predict customer retention. You
may use logistic regression to predict the variable "retained." You can use any
combination of the independent variables available in the data to obtain a model
with the best predictive ability and usability. You are free to use different
transformations and combinations of the independent variables. Be aware that
there is no “magic bullet” to finding the ideal model. You will have to go through
multiple iterations.

In [30]:
print(logreg)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)


In [31]:
print("The accuracy of our Logistic Regression model is " + str(logreg.score(X_train, y_train)*100) + "%")

The accuracy of our Logistic Regression model is 94.18206986401759%


In [32]:
print("The coefficients of our logreg model are " + str(logreg.coef_))

The coefficients of our logreg model are [[ 0.21207411  0.00579462  0.01315318 -0.00355594 -0.57737349  0.38339248
   0.74983522  0.80515318 -1.14030893 -0.03492538 -0.10563622]]


In [33]:
print("The intercept of our logreg model is " + str(logreg.intercept_))

The intercept of our logreg model is [-1.14030893]


### Question 2
Once you obtain the best model that you can find, predict retention in the test
data. You will use the coefficients obtained from the model estimated using the
train data to do this. Name this predicted value "pretain."

In [34]:
print(classification_report(y_test, y_pred))

             precision    recall  f1-score   support

          0       0.99      0.69      0.81      1286
          1       0.93      1.00      0.96      4933

avg / total       0.94      0.93      0.93      6219



### Question 3
Calculate the hit rate. This can be calculated as % of matches between the value
of pretain and retained in the train data.

In [35]:
print(confusion_matrix(y_test, y_pred))

[[ 887  399]
 [   8 4925]]


### Question 4
Be prepared to present your results in class. The team with the best model, as
judged by your peers, will win "bragging rights."